# Querying the atx_osm database

In [1]:
import sqlite3
from pprint import pprint

In [2]:
sqlite_file = "atx_osm.db"

In [3]:
conn = sqlite3.connect(sqlite_file)

In [4]:
cur = conn.cursor()

## List of cities

Cities from nodes_tags table

In [12]:
cur.execute("SELECT value, count(*) as count FROM nodes_tags \
             WHERE key = 'city' \
             GROUP BY value \
             ORDER BY count DESC")

In [13]:
cur.fetchall()

[(u'Austin', 926),
 (u'Round Rock', 31),
 (u'Kyle', 27),
 (u'Pflugerville', 19),
 (u'Cedar Park', 13),
 (u'Buda', 12),
 (u'Bastrop', 5),
 (u'Dripping Springs', 5),
 (u'Georgetown', 5),
 (u'Leander', 4),
 (u'West Lake Hills', 4),
 (u'Wimberley', 3),
 (u'Del Valle', 2),
 (u'Elgin', 2),
 (u'Lago Vista', 2),
 (u'Lakeway', 2),
 (u'Spicewood', 2),
 (u'Taylor', 2),
 (u'Bee Cave', 1),
 (u'Cedar Creek', 1),
 (u'Creedmoor', 1),
 (u'Hutto', 1),
 (u'Manchaca', 1),
 (u'Manor', 1),
 (u'Maxwell', 1),
 (u'Smithville', 1)]

Cities from ways_tags table:

In [14]:
cur.execute("SELECT value, count(*) as count FROM ways_tags \
             WHERE key = 'city' \
             GROUP BY value \
             ORDER BY count DESC")

In [15]:
cur.fetchall()

[(u'Austin', 2142),
 (u'Round Rock', 82),
 (u'Kyle', 37),
 (u'Cedar Park', 30),
 (u'Leander', 29),
 (u'Pflugerville', 18),
 (u'Buda', 14),
 (u'Georgetown', 12),
 (u'Dripping Springs', 8),
 (u'West Lake Hills', 8),
 (u'Elgin', 7),
 (u'Lakeway', 7),
 (u'Bee Cave', 5),
 (u'Taylor', 5),
 (u'Wimberley', 5),
 (u'Bastrop', 4),
 (u'Manor', 4),
 (u'Del Valle', 3),
 (u'Manchaca', 3),
 (u'Cedar Creek', 2),
 (u'Hutto', 2),
 (u'San Marcos', 2),
 (u'Sunset Valley', 2),
 (u'Webberville', 2),
 (u'Creedmoor', 1),
 (u'Driftwood', 1),
 (u'Jonestown', 1),
 (u'Lost Pines', 1),
 (u'Manchacha', 1),
 (u'Spicewood', 1)]

Union of nodes and ways tags:

In [19]:
cur.execute("SELECT tags.value, COUNT(*) as count\
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags \
             WHERE tags.key = 'city' \
             GROUP BY tags.value \
             ORDER By count DESC")

In [20]:
cities = cur.fetchall()
cities

[(u'Austin', 3068),
 (u'Round Rock', 113),
 (u'Kyle', 64),
 (u'Cedar Park', 43),
 (u'Pflugerville', 37),
 (u'Leander', 33),
 (u'Buda', 26),
 (u'Georgetown', 17),
 (u'Dripping Springs', 13),
 (u'West Lake Hills', 12),
 (u'Bastrop', 9),
 (u'Elgin', 9),
 (u'Lakeway', 9),
 (u'Wimberley', 8),
 (u'Taylor', 7),
 (u'Bee Cave', 6),
 (u'Del Valle', 5),
 (u'Manor', 5),
 (u'Manchaca', 4),
 (u'Cedar Creek', 3),
 (u'Hutto', 3),
 (u'Spicewood', 3),
 (u'Creedmoor', 2),
 (u'Lago Vista', 2),
 (u'San Marcos', 2),
 (u'Sunset Valley', 2),
 (u'Webberville', 2),
 (u'Driftwood', 1),
 (u'Jonestown', 1),
 (u'Lost Pines', 1),
 (u'Manchacha', 1),
 (u'Maxwell', 1),
 (u'Smithville', 1)]

## Postcode

In [21]:
cur.execute("SELECT tags.value, COUNT(*) as count\
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags \
             WHERE tags.key = 'postcode' \
             GROUP BY tags.value \
             ORDER By count DESC")

In [22]:
postcode = cur.fetchall()
postcode

[(u'78645', 10882),
 (u'78734', 5606),
 (u'78653', 3543),
 (u'78660', 3516),
 (u'78669', 3189),
 (u'78641', 2863),
 (u'78704', 2488),
 (u'78746', 2446),
 (u'78759', 2092),
 (u'78738', 1939),
 (u'78748', 1915),
 (u'78723', 1896),
 (u'78703', 1845),
 (u'78617', 1829),
 (u'78701', 1726),
 (u'78724', 1680),
 (u'78731', 1680),
 (u'78745', 1592),
 (u'78744', 1525),
 (u'78758', 1419),
 (u'78620', 1323),
 (u'78732', 1307),
 (u'78754', 1289),
 (u'78741', 1286),
 (u'78747', 1277),
 (u'78705', 1237),
 (u'78750', 1212),
 (u'78733', 1199),
 (u'78757', 1160),
 (u'78735', 1067),
 (u'78737', 1053),
 (u'78753', 1040),
 (u'78702', 1027),
 (u'78725', 1027),
 (u'78736', 1009),
 (u'78730', 938),
 (u'78749', 833),
 (u'78751', 813),
 (u'78728', 787),
 (u'78739', 786),
 (u'78721', 756),
 (u'78717', 716),
 (u'78652', 712),
 (u'78610', 685),
 (u'78621', 645),
 (u'78752', 610),
 (u'78729', 599),
 (u'78727', 560),
 (u'78756', 534),
 (u'78719', 475),
 (u'78613', 431),
 (u'78654', 412),
 (u'78726', 392),
 (u'78664'

We can see that there are three "None" values. Determination of the missing postcodes:

In [24]:
cur.execute("SELECT * \
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags \
             WHERE tags.key = 'postcode' \
             AND tags.value = 'None'")

In [25]:
missing_postcodes = cur.fetchall()

In [26]:
missing_postcodes

[(2152207067, u'postcode', u'None', u'addr'),
 (247506590, u'postcode', u'None', u'addr'),
 (383791236, u'postcode', u'None', u'addr')]

##### Looking for info on id 2152207067:

In [44]:
cur.execute("SELECT tags.value \
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags \
             WHERE tags.key = 'city'\
             AND tags.id = 2152207067;")

In [45]:
missing_postcode1 = cur.fetchall()
missing_postcode1

[]

In [47]:
cur.execute("SELECT * \
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags\
             WHERE tags.id = 2152207067")
missing_postcode1_info = cur.fetchall()
missing_postcode1_info

[(2152207067, u'name', u'Nyle Maxwell - Taylor', u'regular'),
 (2152207067, u'shop', u'car', u'regular'),
 (2152207067, u'website', u'www.nylemaxwellcjd.com', u'regular'),
 (2152207067, u'street', u'United States Highway 79', u'addr'),
 (2152207067, u'postcode', u'None', u'addr')]

The missing info for id 2152207067 is then a car dealership address. Searching on the internet for www.nylemaxwellcjd.com, we find that the address is:

14150 UNITED STATES HIGHWAY 79 TAYLOR,	TX	76574

##### Determining the postcode for id 247506590:

In [48]:
cur.execute("SELECT * \
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags\
             WHERE tags.id = 247506590")
missing_postcode2_info = cur.fetchall()
missing_postcode2_info

[(247506590, u'name', u'Concordia University Texas', u'regular'),
 (247506590, u'phone', u'512-313-3000', u'regular'),
 (247506590, u'amenity', u'university', u'regular'),
 (247506590, u'website', u'http://www.concordia.edu', u'regular'),
 (247506590, u'operator', u'Concordia University System', u'regular'),
 (247506590, u'city', u'Austin', u'addr'),
 (247506590, u'wikipedia', u'en:Concordia University Texas', u'regular'),
 (247506590, u'street', u'Concordia University Drive', u'addr'),
 (247506590, u'postcode', u'None', u'addr'),
 (247506590, u'housenumber', u'11400', u'addr')]

Searching for the address shown in the info above using Google Maps, we now know that the postcode is 78726.

https://goo.gl/maps/Bmh7gNoRLGu

##### Determination of postcode for id 383791236:

In [49]:
cur.execute("SELECT * \
             FROM (SELECT * FROM nodes_tags\
             UNION ALL \
             SELECT * FROM ways_tags) tags\
             WHERE tags.id = 383791236")
missing_postcode3_info = cur.fetchall()
missing_postcode3_info

[(383791236, u'building', u'house', u'regular'),
 (383791236, u'city', u'Austin', u'addr'),
 (383791236, u'street', u'Elderberry Road', u'addr'),
 (383791236, u'postcode', u'None', u'addr'),
 (383791236, u'housenumber', u'282', u'addr')]

Searching for the address above on Google maps yield a postcode of 78745.

## Number of nodes:

In [50]:
cur.execute("SELECT COUNT(*) FROM nodes")
nodes = cur.fetchall()
nodes

[(6356394,)]

This value is the same as the one obtained from the exploration of dataset using xml.etree.cElementTree module of Python (see p3_wrangle_openstreetmap_1.ipynb, High Level Tags)

## Number of ways:

In [52]:
cur.execute("SELECT COUNT(*) FROM ways")
ways = cur.fetchall()
ways

[(666390,)]

This is also the same number obtained from the the ElementTree module in Python (p3_wrangle_openstreetmap_1.ipynb, High Level Tags).

## Number of unique users

In [53]:
cur.execute("SELECT COUNT(DISTINCT(e.uid))\
             FROM (SELECT uid from nodes UNION ALL SELECT uid FROM ways) e")
users = cur.fetchall()
users

[(1146,)]

This number is lower than the one obtained using the ElementTree module in the exploration of the xml osm (1155 users, p3_wrangle_openstreetmap_1.ipynb, Exploring Users). This might be because when the csv files were created, the key values with problematic characters were removed, along with the rest of the record containing that value.

## Top 10 contributing users

In [54]:
cur.execute("SELECT e.user, COUNT(*) as num \
             FROM (SELECT user FROM nodes UNION ALL SELECT user FROM ways) e \
             GROUP BY e.user \
             ORDER BY num DESC \
             LIMIT 10")

In [55]:
top10contributors = cur.fetchall()
top10contributors

[(u'patisilva_atxbuildings', 2743705),
 (u'ccjjmartin_atxbuildings', 1300514),
 (u'ccjjmartin__atxbuildings', 940070),
 (u'wilsaj_atxbuildings', 359124),
 (u'jseppi_atxbuildings', 300983),
 (u'woodpeck_fixbot', 223425),
 (u'kkt_atxbuildings', 157847),
 (u'lyzidiamond_atxbuildings', 156383),
 (u'richlv', 50212),
 (u'johnclary_axtbuildings', 48232)]

Contributions of users:

In [56]:
contributions = cur.execute("SELECT e.user, COUNT(*) as num \
                             FROM (SELECT user FROM nodes UNION ALL SELECT user FROM ways) e \
                             GROUP BY e.user \
                             ORDER BY num DESC").fetchall()

### Using pandas:

In [57]:
import pandas as pd

In [59]:
contributions_df = pd.DataFrame(contributions)
contributions_df.head(10)

0        1
0    patisilva_atxbuildings  2743705
1   ccjjmartin_atxbuildings  1300514
2  ccjjmartin__atxbuildings   940070
3       wilsaj_atxbuildings   359124
4       jseppi_atxbuildings   300983
5           woodpeck_fixbot   223425
6          kkt_atxbuildings   157847
7  lyzidiamond_atxbuildings   156383
8                    richlv    50212
9    johnclary_axtbuildings    48232

In [63]:
contributions_df['users'] = contributions_df[0]

In [64]:
contributions_df['count'] = contributions_df[1]

In [65]:
del contributions_df[0]

In [66]:
del contributions_df[1]

In [67]:
contributions_df.head(10)

users    count
0    patisilva_atxbuildings  2743705
1   ccjjmartin_atxbuildings  1300514
2  ccjjmartin__atxbuildings   940070
3       wilsaj_atxbuildings   359124
4       jseppi_atxbuildings   300983
5           woodpeck_fixbot   223425
6          kkt_atxbuildings   157847
7  lyzidiamond_atxbuildings   156383
8                    richlv    50212
9    johnclary_axtbuildings    48232

## Number of users having only one contribution:

Using pandas instead of sqlite database

In [74]:
len(contributions_df[contributions_df['count'] == 1])

256